lib/Analysis/Utility.cpp

```diff
@@ -136,47 +136,46 @@ bool ReduceOpHelper::isWarpSynchronous() {
          (triton::gpu::getWarpsPerCTA(argsLayout)[axis] == 1);
 }
 
-SmallVector<SmallVector<unsigned>> ReduceOpHelper::getScratchConfigsFast() {
-  SmallVector<SmallVector<unsigned>> smemShapes(3);
+SmallVector<unsigned> ReduceOpHelper::getScratchConfigsFast() {
+  SmallVector<unsigned> smemShape;
 
   auto argLayout = getSrcLayout();
   auto argLayoutMma = argLayout.dyn_cast<triton::gpu::MmaEncodingAttr>();
 
   // that case doesn't need inter-warp communication
   if (isWarpSynchronous())
-    return {{0, 0}, {0, 0}};
+    return {0, 0};
 
   /// shared memory block0
-  smemShapes[0] = convertType<unsigned>(getSrcShape());
-  smemShapes[0][axis] = getInterWarpSize();
+  smemShape = convertType<unsigned>(getSrcShape());
+  smemShape[axis] = getInterWarpSize();
 
-  /// FIXME(Qingyi): This size is actually larger than required.
-  /// shared memory block1:
-  auto mod = op->getParentOfType<ModuleOp>();
-  unsigned numWarps = triton::gpu::TritonGPUDialect::getNumWarps(mod);
-  unsigned threadsPerWarp =
-      triton::gpu::TritonGPUDialect::getThreadsPerWarp(mod);
-  smemShapes[1].push_back(numWarps * threadsPerWarp);
+  return smemShape;
+}
 
-  return smemShapes;
+Type ReduceOpHelper::getLargestSrcElementType() {
+  Type elemTy = srcElementTypes[0];
+  for (const auto &ty : srcElementTypes) {
+    if (elemTy.getIntOrFloatBitWidth() < ty.getIntOrFloatBitWidth()) {
+      elemTy = ty;
+    }
+  }
+  return elemTy;
 }
 
 unsigned ReduceOpHelper::getScratchSizeInBytes() {
   unsigned elems = 0;
   if (isFastReduction()) {
-    auto smemShapes = getScratchConfigsFast();
-    for (const auto &smemShape : smemShapes)
-      elems = std::max(elems, product<unsigned>(smemShape));
+    auto smemShape = getScratchConfigsFast();
+    elems = product<unsigned>(smemShape);
   } else {
     auto smemShape = getScratchConfigBasic();
     elems = product<unsigned>(smemShape);
   }
 
-  unsigned bytesPerElem = 0;
-  for (const auto &ty : srcElementTypes) {
-    bytesPerElem += ceil<unsigned>(ty.getIntOrFloatBitWidth(), 8);
-  }
-  return bytesPerElem * elems;
+  unsigned bytesPerElem =
+      ceil<unsigned>(getLargestSrcElementType().getIntOrFloatBitWidth(), 8);
+  return (bytesPerElem * srcElementTypes.size()) * elems;
```

Comment:
Say I'm doing argmax on a float64 tensor, would this will use 16 bytes per element instead of 12?

Reply:
That's right

Reply:
Otherwise, if your first operand is 2 bytes and your second operand is 4 bytes, you can run into misaligned addresses when reading the second operand from smem

Reply:
I see what you mean about alignment issues, but using extra shared memory seems bad. I believe the optimal way to sort the shared memory by descending alignment so you would have `smemBase = {4 * elems, 0}` in your example.

Reply:
Hmm, true, there's no reason for the operands in smem to be in the same order as their index to the op. This is what you meant right ?

Reply:
Yes exactly. So a more complex example, if we had (int8, int32, int16, int64) we would store them in from largest to smallest  (int64, int32, int16, int8).

Reply:
yea that makes sense, thanks for the suggestion!